File: problems/629_k_inverse_pairs_array/Solution.java
class Solution {
    public int kInversePairs(int n, int k) {
        int MOD = 1000000007;
        int[] dp = new int[k + 1];
        dp[0] = 1;
        for (int i = 2; i <= n; i++) {
            int[] newDp = new int[k + 1];
            long[] prefix = new long[k + 2];
            for (int j = 0; j <= k; j++) prefix[j + 1] = (prefix[j] + dp[j]) % MOD;
            for (int j = 0; j <= k; j++) {
                newDp[j] = (int)((prefix[j + 1] - prefix[Math.max(0, j - i + 1)] + MOD) % MOD);
            }
            dp = newDp;
        }
        return dp[k];
    }
}

File: problems/629_k_inverse_pairs_array/solution.py
class Solution:
    def kInversePairs(self, n: int, k: int) -> int:
        MOD = 10**9 + 7
        dp = [0] * (k + 1)
        dp[0] = 1
        for i in range(2, n + 1):
            new_dp = [0] * (k + 1)
            prefix = [0] * (k + 2)
            for j in range(k + 1):
                prefix[j + 1] = (prefix[j] + dp[j]) % MOD
            for j in range(k + 1):
                new_dp[j] = (prefix[j + 1] - prefix[max(0, j - i + 1)]) % MOD
            dp = new_dp
        return dp[k] % MOD
